flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManagerTest.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+/**
+ * Tests for the checkpoint failure manager.
+ */
+public class CheckpointFailureManagerTest extends TestLogger {
+
+	@Test
+	public void testContinuousFailure() {
```

Comment:
Do we need to add a test case for concurrent checkpoint?

Reply:
I think it's not necessary. The counter is `AtomicInteger` and we tolerance concurrent checkpoint and concurrent counting.

Reply:
For concurrent checkpoints, such as chk-1 and chk-2, we may get chk-2 success and then receive chk-1 failure (add a test now and may change the expected result after the change of 12514), in my opinion, we should add a test to guarantee the result is what we want.

Reply:
I know what you mean, please see this [discussion section](https://github.com/apache/flink/pull/8322#discussion_r280328078). Currently, the counting mechanism is not based on checkpoint order. It's a temporal solution, we will give a better solution in the future. Its jira issue id is FLINK-12514.